src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/AnyOS/ManagedPal.Asn.cs

```diff
@@ -143,7 +143,18 @@ public override Oid GetEncodedMessageType(byte[] encodedMessage)
                 reader.GetEncodedValue(),
                 AsnEncodingRules.BER);
 
-            return new Oid(contentInfo.ContentType);
+            switch (contentInfo.ContentType)
+            {
+                case Oids.Pkcs7Data:
+                case Oids.Pkcs7Signed:
+                case Oids.Pkcs7Enveloped:
+                case Oids.Pkcs7SignedEnveloped:
+                case Oids.Pkcs7Hashed:
+                case Oids.Pkcs7Encrypted:
+                    return new Oid(contentInfo.ContentType);
+            }
+
+            throw new CryptographicException(SR.Cryptography_Cms_InvalidMessageType);
```

Comment:
This change cannot be merged without tests exercising the new exceptions (showing that they were exceptions under NetFX)

Reply:
I can port the Mono tests, or write new ones if necessary. The Windows specific PAL in CoreFX already has the same checks.

Reply:
The code in Windows PAL: https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.cs#L109
Mono unit test:https://github.com/mono/mono/blob/master/mcs/class/System.Security/Test/System.Security.Cryptography.Pkcs/ContentInfoTest.cs#L108

The list of allowed OIDs was taken from the Windows CoreFX PAL code. On .NET Framework the value "SignedAndEnvelopedData" also throws an exception. It was obsolete since the original specification of CMS (RFC 2630) afaik.

Reply:
Personally, I'd prefer new tests.  ("Mono-inspired" is fine, mainly I'm just bothered by the double-copyright header stuff)

The comment is about all the exceptions, not just this one.  I thought I had preserved NetFX behavior, but apparently on this one I just had a "oh, clearly this throws if it's corrupt, and otherwise returns the OID, which is from an unrestricted set" feeling.  Having a test which codifies the exception is good.

Then, for CmsSigner.set_SignerIdentifierType, it looks like NetFX has very tight restrictions (https://referencesource.microsoft.com/#System.Security/system/security/cryptography/pkcs/pkcs7signer.cs,81), so apparently a test for that should throw when assigning any of {-1, 0, 4} cast as the enum type.  (And it looks like both Mono and I forgot to test out of range on CmsSigner.IncludeOption).

--

Since our general rule is that things can get "less Exceptiony" but not "more Exceptiony" the tests for the new throw statements just makes sure that we're not being more restrictive than NetFx (and, for PALled types, than other platforms).

Reply:
Fine, I will write new unit tests for the behaviour. I tried to restrict the patch to behaviour that is explicitly stated by the documentation (and independently tested by Mono tests) or already enforced by the Windows PAL code.

Reply:
I committed couple of simple unit tests for the exceptions. I didn't add the checks to CmsSigner.IncludeOption yet, trying to keep the patch small.